agents/wiregrid_encoder/wiregrid_encoder.py

```diff
@@ -0,0 +1,356 @@
+COUNTER_INFO_LENGTH = 100
+COUNTS_ON_BELT = 52000
+REFERENCE_COUNT_MAX = 2 << 15  # > that of belt on wiregrid (=nominal 52000)
+SLEEP = 0.1
+
+
+def count2time(counts, t_offset=0.):
+    t_array = np.array(counts, dtype=float) - counts[0]
+    t_array *= 5.e-9
+    t_array += t_offset
+    return t_array.tolist()
+
+
+class WiregridEncoderAgent:
+    """ Agent to record the wiregrid rotary-encoder data.
+    The encoder signal and IRIG timing signal is read
+    by a BeagleBoneBlack (BBB).
+    The BBB sends the data to this PC via Ethernet.
+
+    Args:
+        bbport(int): Port number of the PC
+                     determined in the script running in the BBB.
+    """
+
+    def __init__(self, agent_obj, bbport=50007):
+
+        self.agent: ocs_agent.OCSAgent = agent_obj
+        self.log = agent_obj.log
+        self.lock = TimeoutLock()
+
+        self.initialized = False
+        self.take_data = False
+
+        self.bbport = bbport
+
+        self.rising_edge_count = 0
+        self.irig_time = 0
+
+        agg_params = {'frame_length': 60}
+        self.agent.register_feed(
+            'wgencoder_rough', record=True,
+            agg_params=agg_params, buffer_time=0.5)
+
+        agg_params = {'frame_length': 60, 'exclude_influx': True}
+        self.agent.register_feed(
+            'wgencoder_full', record=True, agg_params=agg_params)
+
+        self.parser = EncoderParser(beaglebone_port=self.bbport)
+
+    def acq(self, session, params=None):
+        """acq()
+
+        **Process** - Run data acquisition.
+
+        Notes:
+            The most recent data collected is stored in session.data in the
+            structure::
+
+                IRIG data case:
+                    >>> response.session['data']
+                    {'fields':
+                        {
+                         'irig_time': computure unix time
+                                      in receiving the IRIG packet,
+                         'rising_edge_count': PRU clock (BBB clock) count,
+                         'edge_diff': Difference of PRU clock
+                                      from the previous IRIG data,
+                         'irig_sec': IRIG second,
+                         'irig_min': IRIG minuite,
+                         'irig_hour': IRIG hour,
+                         'irig_day': IRIG Day,
+                         'irig_year': IRIG Year
+                        }
+                    }
+
+                Encoder data case:
+                    >>> response.session['data']
+                    {'fields':
+                        {
+                         'quadrature' (list):  quadrature encoder signals,
+                         'pru_clock' (list): PRU clock (Beaglebone clock) ,
+                         'reference_degree' (list): Encoder rotation position
+                                                    [deg.],
+                         'error' (list): Encoder error flags
+                        }
+                    }
```

Comment:
Since I'm not familiar with the external script on the beaglebone -- when are these two cases considered? Is it in IRIG mode or Enocder mode? Can both modes be active at once?

If both modes can be active, then these should be combined, else a client expecting the Encoder mode might be the IRIG mode response and not know what to do.

In any case, you should add a timestamp to these so that the client knows if the data is stale or not.

Reply:
I forgot to write timestamp in the docstring, but timestamp data is already implemented.

The beaglebone sends the IRIG data and Encoder data randomly. 
So, this agent record the IRIG and encoder data in one agent in one agent.
I mean an example of the list of session data is:
[{timestamp, IRIG data}, {timestamp, Encoder data}, {timestamp, Encoder data}, {timestamp, Encoder data}{timestamp, IRIG data}, ...]

Reply:
I merged the irig and encoder data as you suggested.
Please check the new version.

Reply:
Looks good to me, thanks for merging that.